File: 2024/03/solution.py
import re
from pathlib import Path

regex = re.compile(r"(?:(do|don't)\(\)|mul\((\d+),(\d+)\))")


class Solution:

    data: list[str]
    input_file: Path

    def __init__(self, input_file: str | Path):
        self.input_file = Path(input_file)
        self.data = self.input_file.open('r').read()

    def solve_part1(self) -> int:
        return sum(int(a) * int(b) for op, a, b in regex.findall(self.data) if not op)

    def solve_part2(self) -> int:
        result = 0
        enabled = True
        for m in regex.findall(self.data):
            match m[0]:
                case "do":
                    enabled = True
                case "don't":
                    enabled = False
                case _:
                    result += int(m[1]) * int(m[2]) * enabled
        return result


if __name__ == "__main__":
    s = Solution("input.txt")
    print(s.solve_part1())
    print(s.solve_part2())

File: 2024/03/solution.nim
import std/re
import std/strutils

let data: string = readFile("input.txt")
let regex: Regex = re"(?:(do|don't)\(\)|mul\((\d+),(\d+)\))"

proc processData(dampen: bool): int =
  var 
    match_array: array[3, string]
    bounds: (int, int) = (0, 0)
    enabled: bool = true
  while true:
    bounds = data.findBounds(regex, match_array, start = bounds[1])
    if bounds[0] == -1: return
    case match_array[0]:
      of "do":
        enabled = true
      of "don't":
        enabled = not dampen
      else:
        if enabled:
          result += match_array[1].parseInt * match_array[2].parseInt

proc solvePart1(): int =
  return processData(dampen = false)

proc solvePart2(): int = 
  return processData(dampen = true)

echo solvePart1()
echo solvePart2()